### PR TITLE
boards: arm: reel_board: add spi cs gpio

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
Add SPI CS GPIO line to reel_board device tree. This is needed to make the board work with Arduino SPI compatible shields.

Tested on reel_board (1507.1) and reel_board_v2 (1507.2).

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

### CODE REVIEW

File: boards/arm/reel_board/dts/reel_board.dtsi [hunk 1]
```diff
@@ -159,6 +159,7 @@ arduino_spi: &spi3 {
 	sck-pin = <47>;
 	miso-pin = <46>;
 	mosi-pin = <45>;
+	cs-gpios = <&gpio1 12 0>;
```
Comment (jfischer-phytec-iot) [inline boards/arm/reel_board/dts/reel_board.dtsi]: I do not understand why it is necessary, we do not define it at board level. Also it does not have to be the pin D10 on the shield.

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: boards/arm/reel_board/dts/reel_board.dtsi (hunk 1 above)
Comment (henrikbrixandersen) [reply boards/arm/reel_board/dts/reel_board.dtsi]: Quote a few boards do define it at the board level.

The problem with defining it on the shield level is when the board used has on-board SPI devices on the same SPI controller. Simply adding `cs-gpios = <&arduino_header 16 0>;` to the shield dts will overwrite any `cs-gpios` specified in the board dts - and the board may not even rely on GPIOs for CS but use built-in CS lines of the SoC SPI controller.

Regarding the use of D10: This is the SPI SS pin according the the Arduino Uno R3 pinout.

Comment (pabigot) [reply boards/arm/reel_board/dts/reel_board.dtsi]: You could reference it as `<&arduino_header 16 0>` to make that relationship more clear. (Slightly more clear: it's unfortunate that the analog pins were ordered before the digital ones in Zephyr's nexus map, resulting in a 6-position offset for digital pin numbers).

Comment (jfischer-phytec-iot) [reply boards/arm/reel_board/dts/reel_board.dtsi]: Nothing is connected to the spi3 on the reel board, it is for arduino_spi only. As I know SPI_NRFX driver does not support dedicated CS (CSN) (but SPIM hardware). `cs-gpios` for the shield are defined in appropriate dt overlays. Having it here does not hurt, but I do not see any benefit either, rather confusing. @erwango Your opinion?

Comment (erwango) [reply boards/arm/reel_board/dts/reel_board.dtsi]: I also find it confusing. `cs-gpios` only has sense when associate to a slave node.
dtc is able to correctly handle multiple `cs-gpios` \ slaves associations when several overlays are used. 
I don't know how it would play with an orphan `cs-gios`.

**Review by erwango — DISMISSED** [review]

File: boards/arm/reel_board/dts/reel_board.dtsi (hunk 1 above)
Comment (pabigot) [reply boards/arm/reel_board/dts/reel_board.dtsi]: > dtc is able to correctly handle multiple `cs-gpios` \ slaves associations when several overlays are used.

This does not appear to be true, as @henrikbrixandersen noted above: If two shield overlays are used, both of which define distinct `cs-gpios`, the single pin specified in the last shield is the only chip-select defined for the bus, and both devices share it.

Support for defining additional CS pins in shields depends on devicetree source modifications ([proposed by](https://github.com/zephyrproject-rtos/zephyr/issues/12698#issuecomment-457545912) @galak somewhere I can't find anymore) that would allow the set of pins to be extended in overlays, with an associated inferred `reg` based on the number of pins defined.

I see there's recently been a move to using something like:
```
&arduino-spi {
   cs-gpios = <&arduino_header 16 0>; /* D10 */
   device@0 {
     /* ... */ 
   };
};
```
in the shield file.  Though all of them use D10, so I think would make more sense to put that in the board devicetree source for a board that provides an `arduino-spi` alias and requires use of `cs-gpios`.

Comment (erwango) [reply boards/arm/reel_board/dts/reel_board.dtsi]: > > dtc is able to correctly handle multiple `cs-gpios` \ slaves associations when several overlays are used.
> 
> This does not appear to be true, as @henrikbrixandersen noted above: If two shield overlays are used, both of which define distinct `cs-gpios`, the single pin specified in the last shield is the only chip-select defined for the bus, and both devices share it.

If this is not the case, then at least having `cs-gpios` + slave coming from both board dts and overlay works (or used to work, since I remember having verified it).
 
> Support for defining additional CS pins in shields depends on devicetree source modifications ([proposed by](https://github.com/zephyrproject-rtos/zephyr/issues/12698#issuecomment-457545912) @galak somewhere I can't find anymore) that would allow the set of pins to be extended in overlays, with an associated inferred `reg` based on the number of pins defined.
> 

That's when I verfiried it, thanks for the exhumation. 
I think that @galak proposal is the right way. This rather seems like a bug in dtc.

So basically, current change depends on dtc status.  If this proposal is accepted it won't work any longer (in case `cs-gpios` is redefined in overlay).

Comment (pabigot) [reply boards/arm/reel_board/dts/reel_board.dtsi]: > So basically, current change depends on dtc status. If this proposal is accepted it won't work any longer (in case `cs-gpios` is redefined in overlay).

Which proposal: defining `cs-gpios` in the board file, or changing the devicetree syntax to support modifying an existing `cs-gpios`?

I think it's clear that the cleaner approach is to have the board devicetree file define `cs-gpios` to Arduino D10 in the bus node that's aliased by `arduino_spi`.  Then overlays need only specify `reg = <0>;`, which will always refer to the standard Uno CS signal.  This will be less confusing for those using boards where SPI drivers don't use `cs-gpios` instead using a pinmux function to select the right one.

An overlay that needs to use a non-standard first chip-select could continue to override `cs-gpios`, but that would be unnecessary for every existing shield definition.  Similarly an overlay that wants to use `reg = <1>;` would have to have dtc support for adding a second CS signal with whatever devicetree source syntax would do it.

Comment (erwango) [reply boards/arm/reel_board/dts/reel_board.dtsi]: If we go in this direction, what are the impacts on current boards/shields definitions ?

Comment (pabigot) [reply boards/arm/reel_board/dts/reel_board.dtsi]: We could remove `cs-gpios` from the shield overlays when that specifies D10, because it would be redundant.  Which AFAICT is all of the shields that provide `cs-gpios`.  Not all do, so going this way would make those shields work on reel and other Nordic platforms.

Comment (erwango) [reply boards/arm/reel_board/dts/reel_board.dtsi]: what about other boards ?

Comment (pabigot) [reply boards/arm/reel_board/dts/reel_board.dtsi]: If we're agreed this is the right way to go, any that don't already do this should be updated.  There's other cleanup related to making Arduino shields consistent (referencing the nexus map instead of board-specific pins) which is probably on my plate anyway.  I'll plan to do them all in a batch after the backlog of merges gets settled.

Comment (erwango) [reply boards/arm/reel_board/dts/reel_board.dtsi]: ok, fine with me then, thanks

Comment (jfischer-phytec-iot) [reply boards/arm/reel_board/dts/reel_board.dtsi]: @henrikbrixandersen Can you change it as suggested [here](https://github.com/zephyrproject-rtos/zephyr/pull/21610/commits/205ba55bac645f34ceb2cfdc3b0f3283504bb7ee#r361849754), please.

Comment (henrikbrixandersen) [reply boards/arm/reel_board/dts/reel_board.dtsi]: @jfischer-phytec-iot Done.

**Review by pabigot — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]

**Review by anangl — APPROVED** [review]
